### ISSUE TITLE
ZFS hangs NFS

### ISSUE THREAD

**Issue by mickg10** [zfsonlinux/zfs#3991]
Provisionally when accessing a ZFS fs shared via NFS3 from my owncloud instance, sometimes a hard hang of nfsd on the ZFS/NFS server happens:

Just as a guess, maybe it is trying to go through the snapshots somehow ?

Mike

Ubuntu, 3.13.0-66 kernel
spl: 0.6.5.3.1~trusty
zfs*: 0.6.5.3.1~trusty

Relevant export:
/home/media *(rw,no_root_squash,async,no_subtree_check)

[82577.545718] INFO: task nfsd:4069 blocked for more than 120 seconds.
[82577.545737]       Tainted: P           OX 3.13.0-66-generic #108-Ubuntu
[82577.545755] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[82577.545777] nfsd            D ffff88043fc53180     0  4069      2 0x00000000
[82577.545779]  ffff880409327a40 0000000000000046 ffff880410834800 ffff880409327fd8
[82577.545781]  0000000000013180 0000000000013180 ffff880410834800 ffffffffa035c540
[82577.545783]  ffffffffa035c544 ffff880410834800 00000000ffffffff ffffffffa035c548
[82577.545786] Call Trace:
[82577.545790]  [<ffffffff817289b9>] schedule_preempt_disabled+0x29/0x70
[82577.545793]  [<ffffffff8172a825>] __mutex_lock_slowpath+0x135/0x1b0
[82577.545796]  [<ffffffff8172a8bf>] mutex_lock+0x1f/0x2f
[82577.545817]  [<ffffffffa01ff3c4>] zfsctl_lookup_objset+0x34/0x3a0 [zfs]
[82577.545819]  [<ffffffff8172a8b2>] ? mutex_lock+0x12/0x2f
[82577.545837]  [<ffffffffa01c6ca8>] ? rrw_exit+0x68/0x170 [zfs]
[82577.545858]  [<ffffffffa02125f6>] zfs_vget+0x176/0x3e0 [zfs]
[82577.545861]  [<ffffffff8161cabf>] ? skb_copy_datagram_iovec+0x12f/0x2a0
[82577.545882]  [<ffffffffa0230832>] zpl_fh_to_dentry+0x72/0xb0 [zfs]
[82577.545886]  [<ffffffffa06828d0>] ? nfsd_proc_getattr+0xa0/0xa0 [nfsd]
[82577.545889]  [<ffffffff812acd82>] exportfs_decode_fh+0x72/0x2e0
[82577.545892]  [<ffffffff810cd340>] ? getboottime+0x30/0x40
[82577.545900]  [<ffffffffa05fe6b7>] ? cache_check+0xf7/0x3d0 [sunrpc]
[82577.545903]  [<ffffffff810aaf34>] ? __wake_up+0x44/0x50
[82577.545906]  [<ffffffff810c8568>] ? __call_rcu_nocb_enqueue+0xa8/0xc0
[82577.545910]  [<ffffffffa06835f6>] fh_verify+0x2f6/0x5e0 [nfsd]
[82577.545913]  [<ffffffff810cd340>] ? getboottime+0x30/0x40
[82577.545918]  [<ffffffffa0691b59>] nfsd4_putfh+0x49/0x50 [nfsd]
[82577.545923]  [<ffffffffa069355a>] nfsd4_proc_compound+0x56a/0x7d0 [nfsd]
[82577.545927]  [<ffffffffa067fd3b>] nfsd_dispatch+0xbb/0x200 [nfsd]
[82577.545934]  [<ffffffffa05f463d>] svc_process_common+0x46d/0x6d0 [sunrpc]
[82577.545941]  [<ffffffffa05f49a7>] svc_process+0x107/0x170 [sunrpc]
[82577.545945]  [<ffffffffa067f71f>] nfsd+0xbf/0x130 [nfsd]
[82577.545949]  [<ffffffffa067f660>] ? nfsd_destroy+0x80/0x80 [nfsd]
[82577.545951]  [<ffffffff8108b7d2>] kthread+0xd2/0xf0
[82577.545954]  [<ffffffff8108b700>] ? kthread_create_on_node+0x1c0/0x1c0
[82577.545957]  [<ffffffff81734ba8>] ret_from_fork+0x58/0x90
[82577.545959]  [<ffffffff8108b700>] ? kthread_create_on_node+0x1c0/0x1c0

**Comment by behlendorf** [comment]
It appears to be blocked on the `&zfs_snapshot_lock` but from this stack trace it's unclear what process is holding the lock.  Where there any additional stacks on the console when this occurred?  Can it be easily reproduced so we can add some additional debugging?

**Comment by behlendorf** [comment]
Closing as a duplicate of #3997 which was fixed in master with commit f22f900f15c28ebd8b8b764ee6229a7069c74a13
